MericCelik/assignment3/Main.java

```diff
@@ -0,0 +1,106 @@
+import java.util.ArrayList;
+import java.util.HashSet;
+
+/**Solution class that uses Dictionary class.
+ * @author AycaMericCelik*/
+
+public class Main {
+    public static void main(String[] args){
+        String[][] grids = {"aar".split(""), "tcd".split("")};
+        String[] words = {"CAR", "CARD", "CART", "CAT"};
+        Dictionary dict = new Dictionary(words);
+        TrieBasedDictionary dict2  = new TrieBasedDictionary(words);
+
+        System.out.println(findWords(grids, dict));
+        System.out.println(TrieBasedDictionary.findWords(grids,dict2));
+
+    }
+
+    /** Finds all words that can be created using grids. Calls buildAllWords method for each starting grid.
+     * It creates zero-filled visited matrix for each call of buildAllWords method. It passes current grids
+     * coordinates, grids matrix and the dictionary.
+     * @param dict : The dictionary that contains some words and prefixes.
+     * @param grids :The grids matrix to construct words from its elements.
+     * @return set of possible words(no duplicates)
+     * @throws NullPointerException when the parameters are invalid.
+     * */
+    public static HashSet<String> findWords(String[][] grids, Dictionary dict){
+
+        if(grids == null || dict ==null)
+            throw new NullPointerException("Invalid parameter!");
+
+        int rowcount = grids.length;
+        int columncount = grids[0].length;
+
+        HashSet<String> allresults = new HashSet<>();
+        ArrayList<String> result;
+        for(int r=0;r<rowcount;r++){
+            for(int c=0;c<columncount;c++){
+                int[][] visited = new int[rowcount][columncount];
+                result = buildAllWords(grids,r,c,"",visited,dict, new ArrayList<>());
```

Comment:
how about allresults.addAll(buildAllWords(...)) or passing allResults as a parameters to buildAllWords method (if you change the last param from ArrayList<String> to HashSet<String>)?

Reply:
HashSet is a very precious thing to use, I should have used it more frequently in this assignment! :smile: